### PR TITLE
single-file-cli: init at 1.1.49

### LINKED ISSUE #270124: Package request: single-file-cli
**Project description**

A CLI tool for saving a faithful copy of a complete web page in a single HTML file.

**Metadata**

* homepage URL: https://github.com/gildas-lormeau/SingleFile (the entire project)
* source URL: https://github.com/gildas-lormeau/single-file-cli (the CLI tool)
* license: AGPL-3.0
* platforms: Linux, possibly others (not sure)

**Priorities**

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by n8henrie**
This is a CLI tool for saving a faithful copy of a complete web page in a single HTML file

Fixes https://github.com/NixOS/nixpkgs/issues/270124

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by raspher** [comment]
Result of `nixpkgs-review pr 283878` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>single-file-cli</li>
  </ul>
</details>

**Comment by n8henrie** [comment]
Thanks for feedback!

> Builded, however i cannot use this cli.

Any errors? It needs to be passed the path to a browser executable with a flag -- my initial submission used a wrapper to provide one by default, but then would give errors if a user used a different one (due the duplicated argument), so I left it off to allow users to provide their own as needed.

> Can you provide sample command to execute and test it works?

Yes -- in the package itself, as a comment? Or here?

> Could you add some tests to this package itself as well?

Happy to, though this is a web scraper, so I assume the sandbox would make any tests fail.

**Comment by raspher** [comment]
> Happy to, though this is a web scraper, so I assume the sandbox would make any tests fail.

You're right, it probably will fail, my bad.

> > Can you provide sample command to execute and test it works?

> Yes -- in the package itself, as a comment? Or here?

Just an example command to pass inside shell of  `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`

> Any errors? 

I'm testing on remote aarch64 VM and got this 
```shell
$ single-file --back-end jsdom "https://github.com" test.html
Class extends value undefined is not a constructor or null URL: https://github.com
Stack: TypeError: Class extends value undefined is not a constructor or null
    at eval (eval at getPageData (/nix/store/h6h2yb89by5daywr2imrzhg90gkkqngz-single-file-cli-1.1.49/lib/node_modules/single-file-cli/back-ends/jsdom.js:76:6), <anonymous>:1:3173)
    at eval (eval at getPageData (/nix/store/h6h2yb89by5daywr2imrzhg90gkkqngz-single-file-cli-1.1.49/lib/node_modules/single-file-cli/back-ends/jsdom.js:76:6), <anonymous>:1:243)
    at eval (eval at getPageData (/nix/store/h6h2yb89by5daywr2imrzhg90gkkqngz-single-file-cli-1.1.49/lib/node_modules/single-file-cli/back-ends/jsdom.js:76:6), <anonymous>:1:314)
    at eval (<anonymous>)
    at getPageData (/nix/store/h6h2yb89by5daywr2imrzhg90gkkqngz-single-file-cli-1.1.49/lib/node_modules/single-file-cli/back-ends/jsdom.js:76:6)
    at async exports.getPageData (/nix/store/h6h2yb89by5daywr2imrzhg90gkkqngz-single-file-cli-1.1.49/lib/node_modules/single-file-cli/back-ends/jsdom.js:39:10)
    at async capturePage (/nix/store/h6h2yb89by5daywr2imrzhg90gkkqngz-single-file-cli-1.1.49/lib/node_modules/single-file-cli/single-file-cli-api.js:256:20)
    at async runNextTask (/nix/store/h6h2yb89by5daywr2imrzhg90gkkqngz-single-file-cli-1.1.49/lib/node_modules/single-file-cli/single-file-cli-api.js:176:20)
    at async Promise.all (index 0)
    at async capture (/nix/store/h6h2yb89by5daywr2imrzhg90gkkqngz-single-file-cli-1.1.49/lib/node_modules/single-file-cli/single-file-cli-api.js:127:2)
```

i've tried 
```shell
$ single-file --back-end playwright-chromium --browser-headless "https://github.com" test.html
Cannot find module 'playwright'
Require stack:
- /nix/store/h6h2yb89by5daywr2imrzhg90gkkqngz-single-file-cli-1.1.49/lib/node_modules/single-file-cli/back-ends/playwright-chromium.js
- /nix/store/h6h2yb89by5daywr2imrzhg90gkkqngz-single-file-cli-1.1.49/lib/node_modules/single-file-cli/single-file-cli-api.js
- /nix/store/h6h2yb89by5daywr2imrzhg90gkkqngz-single-file-cli-1.1.49/lib/node_modules/single-file-cli/single-file
```

As mentioned, this is environment without desktop/X/wayland

**Comment by n8henrie** [comment]
Sorry for the long response time, thank you again for your time and for testing.

Seems to work for me on both aarch64-linux as well as x86_64-linux.

```console
$ nix-info -m
 - system: `"aarch64-linux"`
 - host os: `Linux 6.1.21, NixOS, 23.11 (Tapir), 23.11.20240202.c3e9c0b`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.19.3`
 - nixpkgs: `/nix/store/7d0pfg1fjznlj4v2wfjl6501asac5qmy-source`
$ nix shell \
    github:n8henrie/nixpkgs/init_single_file_cli#single-file-cli \
    nixpkgs#chromium \
    -c single-file --browser-executable-path=chromium-browser https://n8henrie.com
$ echo $?
0
$ head n8henrie.com\ \(2_6_2024\ 3_32_01\ PM\).html 
<!DOCTYPE html> <html class="wf-opensans-n6-active wf-opensans-n3-active wf-arvo-n7-active wf-arvo-i4-active wf-arvo-n4-active wf-inconsolata-n4-active wf-opensans-i6-active wf-opensans-i3-active wf-active"><!--
 Page saved with SingleFile 
 url: https://n8henrie.com 
 saved date: Tue Feb 06 2024 15:32:01 GMT-0700 (Mountain Standard Time)
--><head><meta charset="utf-8">
  
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <meta name="google-site-verification" content="lOgYEDSRH_etu9NWHYa82k8iJKsCS-zMBXUeGNVyZSA">
```

```console
$ nix-info -m
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.11, NixOS, 23.11 (Tapir), 23.11.20240204.2b41125`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.19.3`
 - channels(root): `""`
 - nixpkgs: `/nix/store/hnr3ckhy6f9qxrwli0mj3vq2rpcaryg7-source`
$ nix shell \
    github:n8henrie/nixpkgs/init_single_file_cli#single-file-cli \
    nixpkgs#chromium \
    -c single-file --browser-executable-path=chromium-browser https://n8henrie.com
$ echo $?
0
$ head n8henrie.com\ \(2_6_2024\ 3_30_29\ PM\).html 
<!DOCTYPE html> <html class="wf-arvo-i4-active wf-arvo-n7-active wf-inconsolata-n4-active wf-arvo-n4-active wf-opensans-i3-active wf-opensans-i6-active wf-opensans-n6-active wf-opensans-n3-active wf-active"><!--
 Page saved with SingleFile 
 url: https://n8henrie.com 
 saved date: Tue Feb 06 2024 15:30:29 GMT-0700 (Mountain Standard Time)
--><head><meta charset="utf-8">
  
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <meta name="google-site-verification" content="lOgYEDSRH_etu9NWHYa82k8iJKsCS-zMBXUeGNVyZSA">
```

**Comment by n8henrie** [comment]
I'm fairly confident I could make a test with localhost-only networking via`python -m http.server --bind 127.0.0.1`. On darwin I think this can be done with the `__darwinallowlocalnetworking` flag, not sure if there is something analogous for linux.

Could also just make it a nixosTest I guess?

**Comment by n8henrie** [comment]
Also working on aarch64-darwin, though I think geckodriver may be the only browser that builds on darwin:

```console
$ nix-info -m
 - system: `"aarch64-darwin"`
 - host os: `Darwin 23.2.0, macOS 14.2.1`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.19.2`
 - channels(n8henrie): `""`
 - channels(root): `""`
 - nixpkgs: `/nix/store/64npyyxn3bb6cxciz12yigldfc2p9fh1-source`
$ nix shell \
    github:n8henrie/nixpkgs/init_single_file_cli#single-file-cli \
    nixpkgs#geckodriver \
    -c single-file \
        --back-end=webdriver-gecko \
        --browser-executable-path=geckodriver \
        https://n8henrie.com
$ echo $?
0
$ head n8henrie.com\ \(2_6_2024\ 3_48_21\ PM\).html
<!DOCTYPE html> <html class="wf-opensans-n6-active wf-opensans-n3-active wf-opensans-i6-active wf-arvo-i4-active wf-arvo-n4-active wf-arvo-n7-active wf-inconsolata-n4-active wf-opensans-i3-active wf-active"><!--
 Page saved with SingleFile
 url: https://n8henrie.com
 saved date: Tue Feb 06 2024 15:48:21 GMT-0700 (Mountain Standard Time)
--><head><meta charset="utf-8">

  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <meta name="google-site-verification" content="lOgYEDSRH_etu9NWHYa82k8iJKsCS-zMBXUeGNVyZSA">
```

**Comment by kirillrdy** [comment]
Result of `nixpkgs-review pr 283878` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>single-file-cli</li>
  </ul>
</details>


here are logs
```
@nix { "action": "setPhase", "phase": "checkPhase" }
Running phase: checkPhase
Serving HTTP on 127.0.0.1 port 8000 (http://127.0.0.1:8000/) ...
127.0.0.1 - - [22/Feb/2024 20:22:54] "GET / HTTP/1.1" 200 -
Browsing context has been discarded URL: http://127.0.0.1:8000
Stack: NoSuchWindowError: Browsing context has been discarded
    at Object.throwDecodedError (/build/source/node_modules/selenium-webdriver/lib/error.js:524:15)
    at parseHttpResponse (/build/source/node_modules/selenium-webdriver/lib/http.js:601:13)
    at Executor.execute (/build/source/node_modules/selenium-webdriver/lib/http.js:529:28)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async thenableWebDriverProxy.execute (/build/source/node_modules/selenium-webdriver/lib/webdriver.js:745:17)
    at async getPageData (/build/source/back-ends/webdriver-gecko.js:115:3)
    at async exports.getPageData (/build/source/back-ends/webdriver-gecko.js:41:10)
    at async capturePage (/build/source/single-file-cli-api.js:256:20)
    at async runNextTask (/build/source/single-file-cli-api.js:176:20)
    at async Promise.all (index 0)

```

```
 nix-info -m
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.16, NixOS, 24.05 (Uakari), 24.05.20240222.278dcd8`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `"nixos"`
 - nixpkgs: `/home/kirillvr/.cache/nixpkgs-review/pr-283878-1/nixpkgs`

```

**Comment by n8henrie** [comment]
@kirillrdy I saw a similar error a few times, seems to be a problem with the test. Wonder if `PYTHONUNBUFFERED` would help, will try.

**Comment by n8henrie** [comment]
Result of `nixpkgs-review pr 283878` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>single-file-cli</li>
  </ul>
</details>

File: pkgs/by-name/si/single-file-cli/package.nix [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    grep -F 'Page saved with SingleFile' 'Directory listing for'*.html
+
+    kill $pid
+    wait
+
+    runHook postCheck
+  '';
+
+  meta = {
```
Comment (kirillrdy) [inline pkgs/by-name/si/single-file-cli/package.nix:48]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#meta-attributes

> meta.mainProgram must be set to the name of the executable which facilitates the primary function or purpose of the package

Comment (n8henrie) [reply pkgs/by-name/si/single-file-cli/package.nix:48]: Thanks!

**Comment by kirillrdy** [comment]
@n8henrie can you update commit messages to match contribution guidelines

**Comment by kirillrdy** [comment]
@n8henrie sorry, can you make two commits

- `maintainers: add n8henrie`
- `single-file-cli: init at 1.1.49`

**Comment by n8henrie** [comment]
No problem, will do shortly

**Comment by kirillrdy** [comment]
FYI 2.0 has been released, but we can deal with that in subsequent PR

**Comment by n8henrie** [comment]
@kirillrdy this is my first "new package" PR -- what is the next step? Do I need to solicit a review?

File: pkgs/by-name/si/single-file-cli/package.nix [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    description = "CLI tool for saving a faithful copy of a complete web page in a single HTML file";
+    homepage = "https://github.com/gildas-lormeau/single-file-cli";
+    license = lib.licenses.agpl3;
```
Comment (kirillrdy) [inline pkgs/by-name/si/single-file-cli/package.nix:51]: error: attribute 'agpl3' missing

**Comment by kirillrdy** [comment]
> @kirillrdy this is my first "new package" PR -- what is the next step? Do I need to solicit a review?

due to large volume of PR's things do get lost a bit, 

as for getting more people to review there is 
there is a discourse https://discourse.nixos.org/t/prs-ready-for-review/3032/3635
and matrix https://app.element.io/#/room/#review-requests:nixos.org

as for this PR, i think I am happy to merge once CI is passing, which it currently isn't

**Comment by n8henrie** [comment]
Interesting, looks like agpl was just removed: https://github.com/NixOS/nixpkgs/commit/82b45bf4547f4d705663e4d2593ac78e24a5d094

Looks like this should be `agpl3Only`: https://github.com/gildas-lormeau/single-file-cli/blob/master/LICENSE

**Comment by kirillrdy** [comment]
Result of `nixpkgs-review pr 283878` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>single-file-cli</li>
  </ul>
</details>

ofborg build fails with same error https://github.com/NixOS/nixpkgs/pull/283878/checks?check_run_id=22999755857

**Comment by n8henrie** [comment]
Huh. I've seen a similar error a few times as well, I'm guessing it's some kind of race condition in setting up the example Python web server, but that's the best idea I had for a simple localhost-only web server to make a request to.

Just adding a `sleep 1` seems pretty cludgy, but is seen elsewhere in nixpkgs so maybe I'll try that.

**Comment by kirillrdy** [comment]
> Huh. I've seen a similar error a few times as well, I'm guessing it's some kind of race condition in setting up the example Python web server, but that's the best idea I had for a simple localhost-only web server to make a request to.
> 
> Just adding a `sleep 1` seems pretty cludgy, but is seen elsewhere in nixpkgs so maybe I'll try that.

instead of adding a single `sleep`, you can do polling using `nc`

otherwise, i would just remove `checkPhase`, and rely on something like `versionTest`

**Comment by n8henrie** [comment]
@ofborg build single-file-cli

**Comment by n8henrie** [comment]
@kirillrdy that error seemed to be specific to using geckodriver (which I used initially because I don't think any other browsers build on darwin), but I didn't find an easy workaround.

Changed to using chromium for the tests and restricting tests to linux -- seems to be working now.

**Comment by kirillrdy** [comment]
@n8henrie thank you !, feel free to tag me for 2.0

**Pull request merged**